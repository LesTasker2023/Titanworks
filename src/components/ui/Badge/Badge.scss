/* Badge Component - BEM Styling with Design Tokens */
.badge {
  display: inline-flex;
  align-items: center;
  border-radius: var(--spacing-1);
  border: 1px solid transparent;
  font-weight: 600;
  transition: var(--transition-fast);
  /* Focus styles */
  &:focus {
    outline: none;
  }
  &:focus-visible {
    outline: 2px solid hsl(var(--ring));
    outline-offset: 2px;
  }
  /* Size variants */
  &--default {
    padding: var(--spacing-1) calc(var(--spacing-2) + var(--spacing-1));
    font-size: var(--text-xs);
  }
  &--sm {
    padding: var(--spacing-1) var(--spacing-2);
    font-size: var(--text-xs);
  }
  &--lg {
    padding: var(--spacing-2) var(--spacing-3);
    font-size: var(--text-sm);
  }
  /* Color variants */
  &--default {
    background-color: hsl(var(--surface-interactive));
    color: hsl(var(--content-inverse));
    box-shadow: var(--shadow-sm);
    &:hover {
      background-color: hsl(var(--surface-interactive) / 0.8);
    }
  }
  &--secondary {
    background-color: hsl(var(--surface-secondary));
    color: hsl(var(--content-primary));
    &:hover {
      background-color: hsl(var(--surface-secondary) / 0.8);
    }
  }
  &--destructive {
    background-color: hsl(var(--status-error));
    color: hsl(var(--content-inverse));
    box-shadow: var(--shadow-sm);
    &:hover {
      background-color: hsl(var(--status-error) / 0.8);
    }
  }
  &--success {
    background-color: hsl(var(--status-success));
    color: hsl(var(--content-inverse));
    box-shadow: var(--shadow-sm);
    &:hover {
      background-color: hsl(var(--status-success) / 0.8);
    }
  }
  &--warning {
    background-color: hsl(var(--status-warning));
    color: hsl(var(--content-inverse));
    box-shadow: var(--shadow-sm);
    &:hover {
      background-color: hsl(var(--status-warning) / 0.8);
    }
  }
  &--info {
    background-color: hsl(var(--status-info));
    color: hsl(var(--content-inverse));
    box-shadow: var(--shadow-sm);
    &:hover {
      background-color: hsl(var(--status-info) / 0.8);
    }
  }
  &--outline {
    background-color: transparent;
    border-color: hsl(var(--border-default));
    color: hsl(var(--content-primary));
    &:hover {
      background-color: hsl(var(--surface-secondary));
    }
  }
  /* Dot variant */
  &--dot {
    padding-left: var(--spacing-2);
  }
}
.badge__dot {
  margin-right: var(--spacing-2);
  height: var(--spacing-2);
  width: var(--spacing-2);
  border-radius: 50%;
  background-color: currentColor;
  opacity: 0.75;
  animation: badge-pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
}
.badge__remove-button {
  margin-left: var(--spacing-2);
  margin-right: calc(var(--spacing-1) * -1);
  height: var(--spacing-4);
  width: var(--spacing-4);
  border-radius: var(--spacing-1);
  opacity: 0.7;
  background: transparent;
  border: none;
  cursor: pointer;
  transition: var(--transition-fast);
  display: flex;
  align-items: center;
  justify-content: center;
  &:hover {
    opacity: 1;
    transform: scale(1.1);
  }
  &:active {
    transform: scale(0.95);
  }
  &:focus {
    outline: none;
  }
  &:focus-visible {
    outline: 2px solid hsl(var(--ring));
    outline-offset: 2px;
  }
}
.badge__remove-icon {
  height: var(--spacing-3);
  width: var(--spacing-3);
  stroke-width: 2;
}
/* Animations */
@keyframes badge-pulse {
  0%,
  100% {
    opacity: 0.75;
  }
  50% {
    opacity: 0.25;
  }
}
/* Reduced motion support */
@media (prefers-reduced-motion: reduce) {
  .badge {
    transition: none;
    &__dot {
      animation: none;
    }
    &__remove-button {
      transition: none;
      &:hover {
        transform: none;
      }
      &:active {
        transform: none;
      }
    }
  }
}
/* High contrast mode support */
@media (prefers-contrast: high) {
  .badge {
    border-width: 2px;
    &--outline {
      border-color: currentColor;
    }
  }
}
