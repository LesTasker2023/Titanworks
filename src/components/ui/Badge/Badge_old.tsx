import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '@/lib/utils';
import { stripTransientProps } from '@/utils/stripTransientProps';

const badgeVariants = cva(
  'inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2',
  {
    variants: {
      variant: {
        default: 'border-transparent bg-primary text-primary-foreground shadow hover:bg-primary/80',
        secondary:
          'border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80',
        destructive:
          'border-transparent bg-destructive text-destructive-foreground shadow hover:bg-destructive/80',
        success: 'border-transparent bg-accent text-accent-foreground shadow hover:bg-accent/80',
        warning:
          'border-transparent bg-status-warning text-white shadow hover:bg-status-warning/80',
        info: 'border-transparent bg-status-info text-white shadow hover:bg-status-info/80',
        outline: 'text-foreground',
      },
      size: {
        default: 'px-2.5 py-0.5 text-xs',
        sm: 'px-2 py-0.5 text-xs',
        lg: 'px-3 py-1 text-sm',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

export interface BadgeProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof badgeVariants> {
  removable?: boolean;
  onRemove?: () => void;
  dot?: boolean;
}

const Badge = React.forwardRef<HTMLDivElement, BadgeProps>(
  (
    { className, variant, size, removable = false, onRemove, dot = false, children, ...props },
    ref
  ) => {
    return (
      <div
        ref={ref}
        className={cn(badgeVariants({ variant, size }), dot && 'pl-1.5', className)}
        {...stripTransientProps(props)}
      >
        {dot && (
          <span
            className="mr-1.5 h-1.5 w-1.5 rounded-full bg-current opacity-75"
            aria-hidden="true"
          />
        )}
        {children}
        {removable && (
          <button
            type="button"
            className="ml-1.5 -mr-1 h-3.5 w-3.5 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2"
            onClick={onRemove}
            aria-label="Remove badge"
          >
            <svg className="h-3 w-3" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M6 18L18 6M6 6l12 12"
              />
            </svg>
          </button>
        )}
      </div>
    );
  }
);

Badge.displayName = 'Badge';

export { Badge, badgeVariants };
