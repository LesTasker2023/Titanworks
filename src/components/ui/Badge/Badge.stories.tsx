import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { Badge } from './Badge';
import './Badge.scss';

const meta = {
  title: 'UI/Badge',
  component: Badge,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'Enhanced Badge component with removable functionality, status dots, and multiple variants.',
      },
    },
  },
  argTypes: {
    variant: {
      control: 'select',
      options: ['default', 'secondary', 'destructive', 'success', 'warning', 'info', 'outline'],
    },
    size: {
      control: 'select',
      options: ['sm', 'default', 'lg'],
    },
    removable: {
      control: 'boolean',
    },
    dot: {
      control: 'boolean',
    },
    children: {
      control: 'text',
    },
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Badge>;

export default meta;
type Story = StoryObj<typeof meta>;

// 1. All Variants
export const AllVariants: Story = {
  render: () => (
    <div className="flex flex-wrap gap-2">
      <Badge variant="default">Default</Badge>
      <Badge variant="secondary">Secondary</Badge>
      <Badge variant="destructive">Destructive</Badge>
      <Badge variant="success">Success</Badge>
      <Badge variant="warning">Warning</Badge>
      <Badge variant="info">Info</Badge>
      <Badge variant="outline">Outline</Badge>
    </div>
  ),
};

// 2. All Sizes
export const AllSizes: Story = {
  render: () => (
    <div className="flex items-center gap-3">
      <Badge size="sm">Small</Badge>
      <Badge size="default">Default</Badge>
      <Badge size="lg">Large</Badge>
    </div>
  ),
};

// 3. Interactive Example
export const InteractiveExample: Story = {
  args: {
    children: 'Interactive Badge',
    variant: 'default',
    size: 'default',
    removable: false,
    dot: false,
  },
};

// 4. Enhanced Features
export const EnhancedFeatures: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">Removable Badges</h4>
        <div className="flex gap-2">
          <Badge
            variant="default"
            removable
            onRemove={() => () => {
              /*  */
            }}
          >
            Removable
          </Badge>
          <Badge
            variant="success"
            removable
            onRemove={() => () => {
              /*  */
            }}
          >
            Tag
          </Badge>
          <Badge
            variant="warning"
            removable
            onRemove={() => () => {
              /*  */
            }}
          >
            Label
          </Badge>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Status Dots</h4>
        <div className="flex gap-2">
          <Badge dot variant="success">
            Online
          </Badge>
          <Badge dot variant="warning">
            Away
          </Badge>
          <Badge dot variant="destructive">
            Offline
          </Badge>
          <Badge dot variant="info">
            Busy
          </Badge>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Combined Features</h4>
        <div className="flex gap-2">
          <Badge
            dot
            removable
            variant="success"
            onRemove={() => () => {
              /*  */
            }}
          >
            Online
          </Badge>
          <Badge
            dot
            removable
            variant="warning"
            onRemove={() => () => {
              /*  */
            }}
          >
            Away
          </Badge>
        </div>
      </div>
    </div>
  ),
};

// 5. Real-world Usage
export const StatusIndicators: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">User Status</h4>
        <div className="flex gap-2">
          <Badge dot variant="success">
            Active
          </Badge>
          <Badge dot variant="warning">
            Pending
          </Badge>
          <Badge dot variant="destructive">
            Suspended
          </Badge>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Priority Levels</h4>
        <div className="flex gap-2">
          <Badge variant="destructive" size="sm">
            High
          </Badge>
          <Badge variant="warning" size="sm">
            Medium
          </Badge>
          <Badge variant="success" size="sm">
            Low
          </Badge>
        </div>
      </div>
    </div>
  ),
};

// 6. Tags and Labels
export const TagsAndLabels: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">Removable Tags</h4>
        <div className="flex flex-wrap gap-2">
          <Badge
            removable
            variant="outline"
            onRemove={() => () => {
              /*  */
            }}
          >
            React
          </Badge>
          <Badge
            removable
            variant="outline"
            onRemove={() => () => {
              /*  */
            }}
          >
            TypeScript
          </Badge>
          <Badge
            removable
            variant="outline"
            onRemove={() => () => {
              /*  */
            }}
          >
            Next.js
          </Badge>
          <Badge
            removable
            variant="outline"
            onRemove={() => () => {
              /*  */
            }}
          >
            Tailwind CSS
          </Badge>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Category Labels</h4>
        <div className="flex flex-wrap gap-2">
          <Badge variant="secondary">Frontend</Badge>
          <Badge variant="info">Backend</Badge>
          <Badge variant="success">DevOps</Badge>
          <Badge variant="warning">Design</Badge>
        </div>
      </div>
    </div>
  ),
};

// 7. Notification Badges
export const NotificationBadges: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">Count Badges</h4>
        <div className="flex gap-4">
          <div className="relative">
            <button className="p-2 border rounded-md">Messages</button>
            <Badge
              variant="destructive"
              size="sm"
              className="absolute -top-1 -right-1 px-1.5 py-0 text-xs"
            >
              3
            </Badge>
          </div>

          <div className="relative">
            <button className="p-2 border rounded-md">Notifications</button>
            <Badge
              variant="warning"
              size="sm"
              className="absolute -top-1 -right-1 px-1.5 py-0 text-xs"
            >
              12
            </Badge>
          </div>
        </div>
      </div>
    </div>
  ),
};

// 8. Size Comparison
export const SizeComparison: Story = {
  render: () => (
    <div className="space-y-4">
      <div className="flex items-center gap-4">
        <div>
          <Badge size="sm" variant="success">
            Small Success
          </Badge>
        </div>
        <div>
          <Badge size="default" variant="success">
            Default Success
          </Badge>
        </div>
        <div>
          <Badge size="lg" variant="success">
            Large Success
          </Badge>
        </div>
      </div>

      <div className="flex items-center gap-4">
        <div>
          <Badge size="sm" variant="warning" removable>
            Small Tag
          </Badge>
        </div>
        <div>
          <Badge size="default" variant="warning" removable>
            Default Tag
          </Badge>
        </div>
        <div>
          <Badge size="lg" variant="warning" removable>
            Large Tag
          </Badge>
        </div>
      </div>
    </div>
  ),
};

// 9. Accessibility Focus States
export const AccessibilityStates: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">Focusable Badges (Removable)</h4>
        <div className="flex gap-2">
          <Badge
            removable
            variant="default"
            onRemove={() => () => {
              /*  */
            }}
          >
            Focus me with Tab
          </Badge>
          <Badge
            removable
            variant="success"
            onRemove={() => () => {
              /*  */
            }}
          >
            Then focus me
          </Badge>
          <Badge
            removable
            variant="warning"
            onRemove={() => () => {
              /*  */
            }}
          >
            And me too
          </Badge>
        </div>
        <p className="text-xs text-muted-foreground mt-2">
          Use Tab to focus the remove buttons, Enter/Space to activate
        </p>
      </div>
    </div>
  ),
};

// 10. Edge Cases
export const EdgeCases: Story = {
  render: () => (
    <div className="space-y-4">
      <div>
        <h4 className="text-sm font-medium mb-2">Long Text</h4>
        <div className="flex flex-wrap gap-2 max-w-md">
          <Badge>Very Long Badge Text That Might Wrap</Badge>
          <Badge removable>Long Removable Badge Text</Badge>
          <Badge dot>Long Status Badge With Dot Indicator</Badge>
        </div>
      </div>

      <div>
        <h4 className="text-sm font-medium mb-2">Empty and Special Cases</h4>
        <div className="flex gap-2">
          <Badge>1</Badge>
          <Badge dot></Badge>
          <Badge removable>X</Badge>
        </div>
      </div>
    </div>
  ),
};
