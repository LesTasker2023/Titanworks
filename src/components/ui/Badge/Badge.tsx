import * as React from 'react';

import { stripTransientProps } from '@/utils/stripTransientProps';
import './Badge.scss';

export interface BadgeProps extends React.HTMLAttributes<HTMLDivElement> {
  variant?: 'default' | 'secondary' | 'destructive' | 'success' | 'warning' | 'info' | 'outline';
  size?: 'default' | 'sm' | 'lg';
  removable?: boolean;
  onRemove?: () => void;
  dot?: boolean;
}

const Badge = React.forwardRef<HTMLDivElement, BadgeProps>(
  (
    {
      className,
      variant = 'default',
      size = 'default',
      removable = false,
      onRemove,
      dot = false,
      children,
      ...props
    },
    ref
  ) => {
    const badgeClassNames = [
      'badge',
      `badge--${variant}`,
      `badge--${size}`,
      dot && 'badge--dot',
      className,
    ]
      .filter(Boolean)
      .join(' ');

    return (
      <div ref={ref} className={badgeClassNames} {...stripTransientProps(props)}>
        {dot && <span className="badge__dot" aria-hidden="true" />}
        {children}
        {removable && (
          <button
            type="button"
            className="badge__remove-button"
            onClick={onRemove}
            aria-label="Remove badge"
          >
            <svg
              className="badge__remove-icon"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M6 18L18 6M6 6l12 12"
              />
            </svg>
          </button>
        )}
      </div>
    );
  }
);

Badge.displayName = 'Badge';

export { Badge };
