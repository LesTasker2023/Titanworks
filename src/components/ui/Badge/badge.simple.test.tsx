import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Badge } from './Badge';

describe('Badge', () => {
  it('renders children correctly', () => {
    render(<Badge>Test Badge</Badge>);
    expect(screen.getByText('Test Badge')).toBeInTheDocument();
  });

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(<Badge>Default</Badge>);
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches all variants snapshot', () => {
      const { container } = render(
        <div data-testid="variants-container">
          <Badge variant="default">Default</Badge>
          <Badge variant="destructive">Destructive</Badge>
          <Badge variant="outline">Outline</Badge>
          <Badge variant="secondary">Secondary</Badge>
        </div>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches all sizes snapshot', () => {
      const { container } = render(
        <div data-testid="sizes-container">
          <Badge size="sm">Small</Badge>
          <Badge size="default">Default</Badge>
          <Badge size="lg">Large</Badge>
        </div>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches removable state snapshot', () => {
      const mockOnRemove = vi.fn();
      const { container } = render(
        <Badge removable onRemove={mockOnRemove}>
          Removable
        </Badge>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches dot indicator snapshot', () => {
      const { container } = render(<Badge dot>With Dot</Badge>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('applies variant classes correctly', () => {
    render(<Badge variant="secondary">Secondary Badge</Badge>);
    const badge = screen.getByText('Secondary Badge');
    expect(badge).toHaveClass('badge--secondary');
  });

  it('supports custom className', () => {
    render(<Badge className="custom-class">Custom Badge</Badge>);
    const badge = screen.getByText('Custom Badge');
    expect(badge).toHaveClass('custom-class');
  });
});
