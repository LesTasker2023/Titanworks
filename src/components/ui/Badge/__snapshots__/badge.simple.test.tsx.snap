// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > Snapshots > matches all sizes snapshot 1`] = `
<div
  data-testid="sizes-container"
>
  <div
    class="badge badge--default badge--sm"
  >
    Small
  </div>
  <div
    class="badge badge--default badge--default"
  >
    Default
  </div>
  <div
    class="badge badge--default badge--lg"
  >
    Large
  </div>
</div>
`;

exports[`Badge > Snapshots > matches all variants snapshot 1`] = `
<div
  data-testid="variants-container"
>
  <div
    class="badge badge--default badge--default"
  >
    Default
  </div>
  <div
    class="badge badge--destructive badge--default"
  >
    Destructive
  </div>
  <div
    class="badge badge--outline badge--default"
  >
    Outline
  </div>
  <div
    class="badge badge--secondary badge--default"
  >
    Secondary
  </div>
</div>
`;

exports[`Badge > Snapshots > matches default snapshot 1`] = `
<div
  class="badge badge--default badge--default"
>
  Default
</div>
`;

exports[`Badge > Snapshots > matches dot indicator snapshot 1`] = `
<div
  class="badge badge--default badge--default badge--dot"
>
  <span
    aria-hidden="true"
    class="badge__dot"
  />
  With Dot
</div>
`;

exports[`Badge > Snapshots > matches removable state snapshot 1`] = `
<div
  class="badge badge--default badge--default"
>
  Removable
  <button
    aria-label="Remove badge"
    class="badge__remove-button"
    type="button"
  >
    <svg
      class="badge__remove-icon"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 18L18 6M6 6l12 12"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;
